db/hints/manager.cc

```diff
@@ -724,15 +714,25 @@ future<> manager::end_point_hints_manager::sender::send_one_hint(lw_shared_ptr<s
             } catch (no_column_mapping& e) {
                 manager_logger.debug("send_hints(): {} at {}: {}", fname, rp, e.what());
                 ++this->shard_stats().discarded;
+            } catch (...) {
+                manager_logger.debug("send_hints(): unexpected error in file {} at {}: {}", fname, rp, std::current_exception());
+                ctx_ptr->on_hint_send_failure(rp);
```

Comment:
I guess the point is to have the "exceptional future ignored" warning appear in the logs, right? Otherwise it should have no effect, as this is a discarded future.

Fixed.

Reply:
After thinking about it some more, I realized that we should be more careful here. If we encounter an unknown exception, we don't know if it is caused by temporary failure or not. 

This line, in its current form, will cause us to fail send_one_file in case of unknown exception. This could cause hints manager to get stuck if the error caused by the hint is persistent.

I'd vote for dropping in this case instead (and will pre-emptively change this PR to do that instead). The worst that could happen is that we lose some hints due to an unknown error that could go away after retry.

Reply:
@piodul No, for the default `catch` clause  you want to treat it as an error and in this case we return exceptional future.

Reply:
> I guess the point is to have the "exceptional future ignored" warning appear in the logs, right? Otherwise it should have no effect, as this is a discarded future.
> 
@piodul No, no need for any extra warning. 
For the default `catch` clause  you want to treat it as an error and in this case we return exceptional future.

Reply:
> I realized that we should be more careful here. If we encounter an unknown exception, we don't know if it is caused by temporary failure or not.
> 
> (The issue mentioned in this comment is not related to returning exceptional future, but is related to this particular piece of code)
> 
> This line, in its current form, will cause us to fail send_one_file in case of unknown exception. This could cause hints manager to get stuck if the error caused by the hint is persistent.
> 
> I'll change it so that we drop the hint instead. It is what we did before this patch, and it is safer this way.

@piodul I'm not sure I can agree with you that before that we are discarding the hint that was handled when we hit the default `catch(...)` clasue today.
As far as I can tell we would fail in sending the current file but the rp for that hint would remain in the rp_set. Therefore would would re-send in the next file_send attempt.

The reason we want to return an exceptional future here is to break sending this file sooner.

The reason for the failure could be any kind of exception in the scope in question, e.g. in `clock::now()`.

Reply:
@vladzcloudius Returning an exceptional future there will not have any more effect than a big warning with a backtrace, printed by the reactor during runtime ("exceptional future ignored"). [Here](https://github.com/scylladb/scylla/blob/517f2c049058a2231853515b43732fadba77017a/db/hints/manager.cc#L687) you can observe that nobody waits for that future - it is explicitly discarded.

In current implementation, no exception handler will be called for that error. Hint position will remain in rps_set, and `segment_replay_failed` won't be set.

It might be confusing because sender::send_one_hint has an exception handler at the top level of `send_one_hint`, but unhandled exceptions in the discarded future won't reach it.

Reply:
> @piodul I'm not sure I can agree with you that we are discarding the hint that was handled when we hit the default catch(...) clasue today.
> As far as I can tell we would fail in sending the current file but the rp for that hint would remain in the rp_set. Therefore we would re-send it in the next file_send attempt.

@vladzcloudius See my comment above for the explanation why (I think) current logic quietly discards the hint in this case.

> The reason for the failure could be any kind of exception in the scope in question, e.g. in clock::now().

My argument for dropping the hint is that in case of `catch (...)` we don't know if the exception will continue to be thrown every time we try to process the same hint. An example would be an exception caused by an incorrectly encoded mutation. Such hint would prevent us from finishing the commitlog segment that contains it.

I think that we should try to add more `catch(some_exception& e)` blocks and choose for each case if we should drop the hint or retry sending the file (not sure if it is in scope of this patch). For errors that we didn't account for, I think it will be safer to drop the hint so that we won't accidentally make hint manager get stuck in a loop, trying to send hints from the same file again and again.

Reply:
> @vladzcloudius Returning an exceptional future there will not have any more effect than a big warning with a backtrace, printed by the reactor during runtime ("exceptional future ignored"). [Here](https://github.com/scylladb/scylla/blob/517f2c049058a2231853515b43732fadba77017a/db/hints/manager.cc#L687) you can observe that nobody waits for that future - it is explicitly discarded.
> 
> In current implementation, no exception handler will be called for that error. Hint position will remain in rps_set, and `segment_replay_failed` won't be set.
> 
> It might be confusing because sender::send_one_hint has an exception handler at the top level of `send_one_hint`, but unhandled exceptions in the discarded future won't reach it.

@piodul I must disagree: returning exceptional future is not going to be ignored since it's going to be eventually caught and handled here: https://github.com/scylladb/scylla/blob/master/db/hints/manager.cc#L732

That's the intent and let's preserve this behaviour. 

I also agree that if there are other special cases like malformed mutation they should be explicitly handled. 

As for the malformed mutation AFAIU this should never happen - in this case I'd expect the corresponding hint segment checksum check to fail which in line should cause the read_log_file() to throw `db::commitlog::segment_error` which is caught and handled in `sender::send_one_file()`.

Reply:
@vladzcloudius Sorry for the long delay. I updated the PR so that in case of an unexpected exception, the hint won't be dropped, but retried later.

> @piodul I must disagree: returning exceptional future is not going to be ignored since it's going to be eventually caught and handled here: https://github.com/scylladb/scylla/blob/master/db/hints/manager.cc#L732

As for the exceptional future issue... the exception handler you linked won't handle it. The hint sending task starts [here](https://github.com/scylladb/scylla/blob/4e7929609086e99afd17799ec4eaa641d871f659/db/hints/manager.cc#L687), and ends [here](https://github.com/scylladb/scylla/blob/4e7929609086e99afd17799ec4eaa641d871f659/db/hints/manager.cc#L731). It's basically:

```C++
(void)with_gate(_file_send_gate, [] { /* stuff */ }).finally([] { /* stuff */ });
```

`with_gate` calls its lambda argument using `futurize_invoke`, so unless the gate is closed, it doesn't throw. A C++ exception thrown inside the lambda argument will be converted into exceptional future. Because the result is discarded with `(void)`, it will be just reported as an exceptional future.

Therefore there is no need for returning `make_exception_future`. A message will be printed similar to this: "send_hints(): unexpected error in file {} at {}: {}", and sending of the hint will be retried later.